test/extended/oauth/requestheaders.go

```diff
@@ -0,0 +1,390 @@
+}
+
+func generateCert(caCert *x509.Certificate, caKey *rsa.PrivateKey, ekus []x509.ExtKeyUsage) (*x509.Certificate, *rsa.PrivateKey) {
+	certTemplate := &x509.Certificate{
+		SerialNumber: big.NewInt(1653),
+		Subject: pkix.Name{
+			Organization: []string{"Testing Org"},
+			Country:      []string{"Faraway"},
+			CommonName:   "testing-client-cert",
+		},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(1, 0, 0),
+		IsCA:                  false,
+		ExtKeyUsage:           ekus,
+		KeyUsage:              x509.KeyUsageDigitalSignature,
+		BasicConstraintsValid: true,
+	}
+
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	pub := &priv.PublicKey
+	certBytes, err := x509.CreateCertificate(rand.Reader, certTemplate, caCert, pub, caKey)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	cert, err := x509.ParseCertificate(certBytes)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	return cert, priv
+}
+
+func waitForNewOAuthConfig(oc *exutil.CLI, configChanged time.Time) {
+	// check that the pods running in openshift-authentication NS already reflect our changes
+	err := wait.PollImmediate(time.Second, 2*time.Minute, func() (bool, error) {
+		pods, err := oc.AdminKubeClient().CoreV1().Pods("openshift-authentication").List(metav1.ListOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		podsReady := true
+		for _, p := range pods.Items {
+			tstamp := p.GetCreationTimestamp()
+			if !tstamp.After(configChanged) {
+				podsReady = false
+			}
+			if podsReady {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	err = wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+		authn, err := oc.AdminConfigClient().ConfigV1().ClusterOperators().Get("authentication", metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		if clusteroperatorhelpers.IsStatusConditionTrue(authn.Status.Conditions, configv1.OperatorProgressing) ||
+			clusteroperatorhelpers.IsStatusConditionTrue(authn.Status.Conditions, configv1.OperatorProgressing) {
+			return false, nil
+		}
+		return true, nil
+	})
+	o.Expect(err).NotTo(o.HaveOccurred())
+}
+
+// returns token or encountered error should it receive Unauthorized or any other error
+func oauthHTTPRequest(oauthBaseURL, endpoint, token string, cert *x509.Certificate, key *rsa.PrivateKey) *http.Response {
+	if (cert == nil) != (key == nil) { // YOU MONSTER!
+		g.Fail("must either specify both key and cert, or neither")
+	}
+
+	req, err := http.NewRequest(http.MethodGet, oauthBaseURL+endpoint, nil)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	if len(token) == 0 {
+		// requesting a token, set user header
+		req.Header.Set("X-Remote-User", "franta")
+	} else {
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+	}
+
+	transport := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true, // we don't care about the server certs here
```

Comment:
I'm not sure the router CA appears in the system bundle but it should not be hard to retrieve it, I'll add that, perhaps there could be cases where even that would fail.